### PR TITLE
Refactored batch API and global architecture

### PR DESCRIPTION
**Pull request by n1k0**
This is now ready for preliminary review.

r=? @leplatrem

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -204,124 +184,89 @@ export default class KintoApi {
...
+      headers: {
+        ...this.optionHeaders,
+        ...headers
       }
     });
-    return results;
+    fn(batch);
+    return this._batchRequests(batch.requests, options)
+      .then((responses) => {
+        if (options.aggregate) {
```
Comment (leplatrem) [inline src/index.js]: nit: `groupByStatus` ?

File: test/api_test.js [hunk 2]
```diff
@@ -374,186 +313,167 @@ describe("Api", () => {
...
-        it("should map create & update requests", () => {
-          expect(requestBody.requests[0]).eql({
-            body: {
-              data: { id: 1, title: "foo" },
-            },
-            headers: { "If-Match": quote(42) },
-            method: "PUT",
-            path: `/${SPV}/buckets/blog/collections/articles/records/1`,
+        // XXX: batch() and requests should allow dealing with the If-Match
+        // header through some "safe" option. Keeping old tests skipped there.
```
Comment (leplatrem) [inline test/api_test.js]: I don't understand this comment.

Comment (n1k0) [inline test/api_test.js]: Forgot to push a commit, fixed by f991e05289398988150417500a4a95d62c02cc71

**Comment by leplatrem** [comment]
I have one concern left :)

The signature of the `createRecord()` method.

Currently, we do that: 
```js
api.batch(batch => {
  batch.createBucket("custom");
  batch.createCollection("blog");
  batch.createRecord("blog", {title: "art1"});
  batch.createRecord("blog", {title: "art2"});
}, {bucket: "custom"})
```

I believe the `collection` should be an option instead (for consistency with bucket):

```js
api.batch(batch => {
  batch.createBucket("custom");
  batch.createCollection("blog");
  batch.createRecord({title: "art1"});
  batch.createRecord({title: "art2"});
}, {bucket: "custom", collection: "blog"})```

**Comment by n1k0** [comment]
> I believe the collection should be an option instead (for consistency with bucket):

I agree. Makes me think we could also define default options for these at the constructor level:

```js
const api = new Api("http://...", {bucket: "custom", collection: "blog"});
api.createRecord({title: "foo"});
```

What do you think?

**Comment by leplatrem** [comment]
> What do you think?

I like it that way. It makes it very convienient to use.

Regarding state about default options for methods, I find it acceptable. But I may not be the most demanding person regarding that matter

**Comment by n1k0** [comment]
Sold then, I'll update the patch accordingly.

**Comment by n1k0** [comment]
I'm just realizing that while there's a `default` bucket, there's no `default` collection. So I think passing the collection name as a required argument still makes sense.

Also, passing a required value as an *option* is highly confusing.

Thoughts @leplatrem?

**Comment by leplatrem** [comment]
>  passing a required value as an option is highly confusing.

True.

r+ then :)

**Pull request merged**